Dev-services and related tests for Oracle

## Description

Follow-up on #15964 : that PR introduces support for Oracle's JDBC driver, but lacks the corresponding dev-services, as well as the corresponding tests (which, crucially, also test that Agroal's `ExceptionSorter` is correctly set).

## Implementation ideas

See `io.quarkus.jdbc.db2.deployment.DevServicesDB2DatasourceTestCase`: basically, we need the same test in the jdbc-oracle extension.

The implementation should be rather similar to what we did for DB2.

Hi, this is a follow-up to the (very welcome!) release of Oracle support in Quarkus native.

There seems to be a bug wherein a custom `reflect-config.json` is ignored whenever the Oracle JDBC support is included.

When included, all of my declared runtime dependencies throw errors such as:

    Caused by: java.lang.InstantiationException: Type `package.for.my.instantiated.ClassName` can not be instantiated reflectively 
    as it does not have a no-parameter constructor or the no-parameter constructor has not been added explicitly to the native image.

I have already declared `package.for.my.instantiated.ClassName` inside `reflect-config.json` and it works fine without the Oracle driver.

## Minimal test case:

The following in `application.properties` works:

      quarkus.native.additional-build-args=\
        -H:ReflectionConfigurationFiles=reflect-config.json

To test, include the following in `src/main/resources/reflect-config.json`, but _without_ the `quarkus-jdbc-oracle`
dependency in the POM:

      [{"name":"foo", "allPublicMethods":true}]

The build fails, as expected, because `foo` is not a known class.

However, when the Oracle dependency is included in the POM (this is the only change):

        <dependency>
            <groupId>io.quarkus</groupId>
            <artifactId>quarkus-jdbc-oracle</artifactId>
        </dependency>

...the build now proceeds. This indicates the reflection config is being ignored. 

I suspect there is a problem merging the native-image Args needed for the Oracle driver with the custom ones 
declared in my project.

I haven't found a workaround yet, so this is stopping us from using Quarkus native with Oracle.

Let me know if this should be a separate bug. 

Thanks very much.

--- a/extensions/jdbc/jdbc-oracle/deployment/pom.xml
+++ b/extensions/jdbc/jdbc-oracle/deployment/pom.xml
@@ -33,6 +33,21 @@
             <groupId>io.quarkus</groupId>
             <artifactId>quarkus-devservices-oracle</artifactId>
         </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-agroal-deployment</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-junit5-internal</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.assertj</groupId>
+            <artifactId>assertj-core</artifactId>
+            <scope>test</scope>
+        </dependency>
     </dependencies>
 
     <build>
@@ -49,6 +64,40 @@
                     </annotationProcessorPaths>
                 </configuration>
             </plugin>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-surefire-plugin</artifactId>
+                <configuration>
+                    <excludes>
+                        <exclude>**/DevServices*TestCase.java</exclude>
+                    </excludes>
+                </configuration>
+            </plugin>
         </plugins>
     </build>
+
+    <profiles>
+        <profile>
+            <id>test-devservices</id>
+            <activation>
+                <property>
+                    <name>test-containers</name>
+                </property>
+            </activation>
+            <build>
+                <plugins>
+                    <plugin>
+                        <groupId>org.apache.maven.plugins</groupId>
+                        <artifactId>maven-surefire-plugin</artifactId>
+                        <configuration>
+                            <excludes>
+                                <!-- force the default, overriding the DevServices exlude -->
+                                <exclude>**/*$*</exclude>
+                            </excludes>
+                        </configuration>
+                    </plugin>
+                </plugins>
+            </build>
+        </profile>
+    </profiles>
 </project>

--- a/extensions/jdbc/jdbc-oracle/deployment/pom.xml
+++ b/extensions/jdbc/jdbc-oracle/deployment/pom.xml
@@ -91,7 +91,7 @@
                         <artifactId>maven-surefire-plugin</artifactId>
                         <configuration>
                             <excludes>
-                                <!-- force the default, overriding the DevServices exlude -->
+                                <!-- force the default, overriding the DevServices exclude -->
                                 <exclude>**/*$*</exclude>
                             </excludes>
                         </configuration>

--- a/extensions/jdbc/jdbc-oracle/deployment/src/test/java/io/quarkus/jdbc/oracle/deployment/DevServicesOracleDatasourceTestCase.java
+++ b/extensions/jdbc/jdbc-oracle/deployment/src/test/java/io/quarkus/jdbc/oracle/deployment/DevServicesOracleDatasourceTestCase.java
@@ -1,0 +1,58 @@
+package io.quarkus.jdbc.oracle.deployment;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.junit.jupiter.api.Assertions.fail;
+
+import java.sql.Connection;
+import java.util.logging.Level;
+import java.util.logging.LogRecord;
+
+import javax.inject.Inject;
+
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.agroal.api.AgroalDataSource;
+import io.agroal.api.configuration.AgroalConnectionPoolConfiguration;
+import io.agroal.api.exceptionsorter.OracleExceptionSorter;
+import io.quarkus.test.QuarkusUnitTest;
+
+public class DevServicesOracleDatasourceTestCase {
+
+    @RegisterExtension
+    static QuarkusUnitTest test = new QuarkusUnitTest()
+            .withEmptyApplication()
+            // Expect no warnings (in particular from Agroal)
+            .setLogRecordPredicate(record -> record.getLevel().intValue() >= Level.WARNING.intValue()
+                    // There are other warnings: JDK8, TestContainers, drivers, ...
+                    // Ignore them: we're only interested in Agroal here.
+                    && record.getMessage().contains("Agroal"))
+            .assertLogRecords(records -> assertThat(records)
+                    // This is just to get meaningful error messages, as LogRecord doesn't have a toString()
+                    .extracting(LogRecord::getMessage)
+                    .isEmpty());
+
+    @Inject
+    AgroalDataSource dataSource;
+
+    @Test
+    public void testDatasource() throws Exception {
+        AgroalConnectionPoolConfiguration configuration = null;
+
+        try {
+            configuration = dataSource.getConfiguration().connectionPoolConfiguration();
+        } catch (NullPointerException e) {
+            // we catch the NPE here as we have a proxycd  and we can't test dataSource directly
+            fail("Datasource should not be null");
+        }
+        assertTrue(configuration.connectionFactoryConfiguration().jdbcUrl().contains("jdbc:oracle:"));
+        assertEquals("quarkus", configuration.connectionFactoryConfiguration().principal().getName());
+        assertEquals(20, configuration.maxSize());
+        assertThat(configuration.exceptionSorter()).isInstanceOf(OracleExceptionSorter.class);
+
+        try (Connection connection = dataSource.getConnection()) {
+        }
+    }
+}

